C#: add System.Memory dependency

jtattermusch opened the pull request:
Also add useful Span-based methods for ByteString

System.Memory dependency is only added for netstandard2.0 target, netstandard1.0 and net45 stay unchanged (we actually could add System.Memory dependency for net45 too, but it seems that most users that will benefit from Span optimizations will already have access to netstandard2.0).

This PR is mostly preparation for future span-based optimizations, but to make at least some use of the new dependency, two utility ByteString methods are being added.

See https://github.com/grpc/grpc-dotnet/issues/30 for more details about the planned serialization/deserialization API.

jtattermusch commented:
@jskeet  WDYT about changing the target from `netstandard1.0` to `netstandard1.1`?

CC @JunTaoLuo @davidfowl @JamesNK

jtattermusch commented:
Looks like this change is mutually exclusive with https://github.com/protocolbuffers/protobuf/pull/5445. I'll reach out to the internal user that requested downgrading LangVersion to see what we can do.

jskeet commented:
I think I'd prefer to *add* a netstandard2.0 target (which comes with a bunch of benefits for 2.0+ users) and only conditionally include System.Memory things.

(Jan, I know a bit more about the LangVersion change - ping me if you need to.)

JunTaoLuo commented on csharp/src/Google.Protobuf/Google.Protobuf.csproj:
```diff
@@ -7,7 +7,7 @@
     <LangVersion>6</LangVersion>
     <Authors>Google Inc.</Authors>
-    <TargetFrameworks>netstandard1.0;net45</TargetFrameworks>
+    <TargetFrameworks>netstandard1.1;net45</TargetFrameworks>
```
This would be considered a breaking change no? Adding another target for netstandard2.0 (or netstandard 1.1 if you want) would probably be better here.

JamesNK replied:
+1

JunTaoLuo commented on csharp/src/Google.Protobuf/Google.Protobuf.csproj:
```diff
@@ -26,11 +26,12 @@
   <ItemGroup>
-    <PackageReference Include="SourceLink.Create.CommandLine" PrivateAssets="All" Version="2.7.6"/>
+    <PackageReference Include="SourceLink.Create.CommandLine" PrivateAssets="All" Version="2.7.6" />
+    <PackageReference Include="System.Memory" Version="4.5.2" />
```
Adding a new dependency is also usually considered a "breaking change". Maybe make this conditional on the target framework?

davidfowl replied:
> Adding a new dependency is also usually considered a "breaking change". Maybe make this conditional on the target framework?

Adding a new dependency isn't a breaking change. Removing one is though. I think we should potentially bump the minor version here.

jtattermusch replied:
Minor version will be bumped because this change won't be available before protobuf v3.8.x is out.

jtattermusch commented:
I'll wait for https://github.com/protocolbuffers/protobuf/pull/5838 to be merged first.

jtattermusch commented:
I added `netstandard2.0` target and added conditional dependency on System.Memory for netstandard2.0 (It would have been possible to add also for net45, but I'm not sure how useful that would be).

JamesNK commented:
FYI I'm halfway through a large PR to add span-based reader and writer for protobuf. Should be able to create a PR in the next day or two.

jtattermusch commented:
@anandolee  @jskeet can we review and merge?

anon17 commented:
I think in the `FromStream` method you can use the "avoid a copy" branch on netstandard2.0, MemoryStream supports GetBuffer there.
